src/Core/Resolvers/SqlMutationEngine.cs

```diff
@@ -1100,37 +1149,199 @@ private void PopulateParamsFromRestRequest(Dictionary<string, object?> parameter
+        ///                         publisher: { name: "Macmillan publishers" },
+        ///                         authors: [{ birthdate: "1997-09-03", name: "Red house authors", author_name: "Dan Brown" }]
+        ///                     },
+        ///                     {
+        ///                         title: "book #2",
+        ///                         reviews: [{ content: "Awesome book." }, { content: "Average book." }],
+        ///                         publisher: { name: "Pearson Education" },
+        ///                         authors: [{ birthdate: "1990-11-04", name: "Penguin Random House", author_name: "William Shakespeare" }]
+        ///                     }])
+        ///                 {
+        ///                     items{
+        ///                         id
+        ///                         title
+        ///                     }
+        ///                 }</example>
+        private void PopulateMutationEntityAndFieldsToAuthorize(
+            Dictionary<string, HashSet<string>> entityToFields,
+            IInputField schema,
+            string entityName,
+            IMiddlewareContext context,
+            object? parameters)
         {
-            string role = string.Empty;
-            if (context.ContextData.TryGetValue(key: AuthorizationResolver.CLIENT_ROLE_HEADER, out object? value) && value is StringValues stringVals)
+            InputObjectType schemaObject = ResolverMiddleware.InputObjectTypeFromIInputField(schema);
+            if (parameters is List<ObjectFieldNode> listOfObjectFieldNode)
             {
-                role = stringVals.ToString();
+                // For the example createbook mutation written above, the object value for `item` is interpreted as a List<ObjectFieldNode> i.e.
+                // all the fields present for item namely- title, reviews, publisher, authors are interpreted as ObjectFieldNode.
+                ProcessObjectFieldNodesForAuthZ(
+                    context: context,
+                    entityName: entityName,
+                    schemaObject: schemaObject,
+                    fieldNodes: listOfObjectFieldNode,
+                    entityTofields: entityToFields);
             }
-
-            if (string.IsNullOrEmpty(role))
+            else if (parameters is List<IValueNode> listOfIValueNode)
             {
-                throw new DataApiBuilderException(
-                    message: "No ClientRoleHeader available to perform authorization.",
-                    statusCode: HttpStatusCode.Unauthorized,
-                    subStatusCode: DataApiBuilderException.SubStatusCodes.AuthorizationCheckFailed);
+                // For the example createbooks mutation written above, the list value for `items` is interpreted as a List<IValueNode>
+                // i.e. items is a list of ObjectValueNode(s).
+                listOfIValueNode.ForEach(iValueNode => PopulateMutationEntityAndFieldsToAuthorize(
```

Comment:
What is the schemaObject for this `if` condition evaluate to? where parameters is List<IValueNode>... Do we really need the schemaObject for this if and the last else if?

Reply:
> What is the schemaObject for this if condition evaluate to? 

This evaluates to the input type of `items` argument without the array modifier, i.e. for createBooks, it will evaluate to `createBookInput` type.

> Do we really need the schemaObject for this if and the last else if?

No, I'll move this evaluation to where it is required.